text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+---
+feature name: Open context provider framwork
+start date: 2020-05-27
+rfc pr: 
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/18
+---
+
+# Summary
+
+The CDK has a super-handy feature in the form of the ContextProvider ( https://docs.aws.amazon.com/cdk/latest/guide/context.html ). These
+ContextProviders are currently hard-coded into the CDK CLI, and additional ContextProviders cannot be added by external sources. This
+proposal is to make the ContextProvider mechanism pluggable so that 3rd parties can create ContextProviders without adding them to the CDK
+repository, and users of the CDK can use these 3rd party ContextProviders.
+
+# Motivation
+
+Development within the CDK Construct library itself benefits greatly from this capability. When a new ContextProvider is needed, then a CDK
+contributor can just create it. Unfortunately, since the providers are hard-coded into the CDK, no 3rd party construct library can create their
+own ContextProvider; other than by contributing it to the CDK, of course. Some ContextProviders may be too specialized for reasonable consideration
+as a CDK contribution.
+
+Instead 3rd party libraries are required to implement workarounds, like CfnCustomResources that resolve during deployment instead of during
+synthesis. This increases the size of the Cfn template generated by the app, increases the time required to deploy the stack, adds unnecessary
+extra resources to the user's account, and makes it possible for a value generated by the CfnCustomResource to differ from one deployment
+to the next.
+
+Unfortunately, these CfnCustomResources required by 3rd party construct libraries may also be a security risk to the user. The lambda that
+backs them persists as long as the stack is deployed, and that lambda is granted permissions into the account that may be possible to leverage
+by an attacker that gains access to the lambda; depending on the lambda's purpose, of course.
+
+# Basic Example
+
+```ts
+// Allow the provider to be an arbitrary string
+const someValues: thirdpartlib.DoSomethingContextReponse = ContextProvider.getValue(scope, {
+  provider: 'thidparty-do-something',
```

Comment:
My thought here was an even MORE open-ended system. How about we do the following instead:

```ts
provider: 'npm:my-provider-package@3.1.2'
```

The CLI can then go and fetch that package (version number required!) and call the context provider using a well-defined protocol: potentially the default export of the module, or a function with a well-known name in there, and some agreed-upon arguments.

This seems like the most open-ended system and has the advantage that the user doesn't ultimately have to go `cdk --plugin`.

Reply:
Of course, what about an environment where `npmjs.com` might not be available? (wink wink nudge nudge)

I guess we need a way to tell `cdk` what `npm install` command to use, so we can subsitute it in select environments?

(Also note that we can't just rely on your project's `package.lock` to bring in the plugin, as not everyone will be running an NPM project. Java/Python/C# users will have a global install of the `cdk`)